utility/session.go

```diff
@@ -1,157 +1,266 @@
-	return s.sessionId
-}
+	if err := sessionHydrator.hydrateSessionID(); err != nil {
+		return nil, fmt.Errorf("failed to hydrate session ID: %w", err)
+	}
 
-func (s *session) GetSessionHeight() int64 {
-	return s.height
-}
+	if err := sessionHydrator.hydrateSessionServicers(); err != nil {
+		return nil, fmt.Errorf("failed to hydrate session servicers: %w", err)
+	}
+
+	if err := sessionHydrator.hydrateSessionFishermen(); err != nil {
+		return nil, fmt.Errorf("failed to hydrate session fishermen: %w", err)
+	}
 
-func (s *session) GetRelayChain() RelayChain {
-	return s.relayChain
+	return sessionHydrator.session, nil
 }
 
-func (s *session) GetGeoZone() GeoZone {
-	return s.geoZone
+type sessionHydrator struct {
+	logger modules.Logger
+
+	// The session being hydrated and returned
+	session *coreTypes.Session
+
+	// The height at which the request is being made to get session information
+	blockHeight int64
+
+	// Caches a readCtx to avoid draining too many connections to the database
+	readCtx modules.PersistenceReadContext
+
+	// A redundant helper that maintains a hex decoded copy of `session.Id` used for session hydration
+	sessionIdBz []byte
 }
 
-func (s *session) GetApplication() *coreTypes.Actor {
-	return s.application
+// hydrateSessionMetadata hydrates the height at which the session started, its number, and the number of blocks per session
+func (s *sessionHydrator) hydrateSessionMetadata() error {
+	numBlocksPerSession, err := s.readCtx.GetIntParam(types.BlocksPerSessionParamName, s.blockHeight)
+	if err != nil {
+		return err
+	}
+	numBlocksAheadOfSession := s.blockHeight % int64(numBlocksPerSession)
+
+	s.session.NumSessionBlocks = int64(numBlocksPerSession)
+	s.session.SessionNumber = int64(s.blockHeight / int64(numBlocksPerSession))
+	s.session.SessionHeight = s.blockHeight - numBlocksAheadOfSession
+	return nil
 }
 
-func (s *session) GetFishermen() []*coreTypes.Actor {
-	return s.fishermen
+// hydrateSessionApplication hydrates the full Application actor based on the address provided
+func (s *sessionHydrator) hydrateSessionApplication(appAddr string) error {
+	// TECHDEBT: We can remove this decoding process once we use `strings` instead of `[]byte` for addresses
+	addr, err := hex.DecodeString(appAddr)
+	if err != nil {
+		return err
+	}
+	s.session.Application, err = s.readCtx.GetActor(coreTypes.ActorType_ACTOR_TYPE_APP, addr, s.session.SessionHeight)
+	return err
 }
 
-func (s *session) GetServicers() []*coreTypes.Actor {
-	return s.servicers
+// validateApplicationSession validates that the application can have a valid session for the provided relay chain and geo zone
+func (s *sessionHydrator) validateApplicationSession() error {
+	app := s.session.Application
+
+	if !slices.Contains(app.Chains, s.session.RelayChain) {
+		return fmt.Errorf("application %s does not stake for relay chain %s", app.Address, s.session.RelayChain)
+	}
+
+	if !(app.PausedHeight == -1 && app.UnstakingHeight == -1) {
```

Comment:
Simplify for readability:

```suggestion
	if app.PausedHeight == -1 || app.UnstakingHeight == -1 {
```h

Reply:
![image](https://user-images.githubusercontent.com/600733/234255840-7614d5eb-2dd2-445e-9266-c7f7b86436d1.png)

Reply:
This broke some of my tests so I had to check

https://twitter.com/olshansky/status/1651680793163603969

<img width="637" alt="Screenshot 2023-04-27 at 1 11 23 PM" src="https://user-images.githubusercontent.com/1892194/234980180-6a566056-03d6-48a7-9cb0-cac77843a89f.png">